File: Stacks/infix postfix perfix converstion/convertion.py
# New implementation

# change logs from the old implementation 

"""
[22 Aug 2022]
1. Full stack implemenation instead of array
2. Added the parenthesis handling for infix to postfix
3. Plain simple approach to OOP based approached
"""

"""
[25 Aug 2022]
1. Added infix to prefix convertion in the driver code
2. Added test expressions to driver code
3. Converted single line comments to multiline comments
4. Changed print statement to return statement for printing converted expression
"""


class Convert:

    def __init__(self) -> None:
        self.top = -1 # default top position of the pointer in the stack
        self.stack = [] # array implementation os stack used for operators handling
        self.output = [] # array for the output expression
        self. precedence = {
            "+" : 1, "-" : 1,
            "*" : 2, "/" : 2,
            "^" : 3
        }
    
    # ===== STACK OPERATIONS =====

    def isEmpty(self) -> bool:
        if self.top == -1:
            return True
        else:
            return False

    def push(self, operator:str) -> None:
        self.top += 1
        self.stack.append(operator)

    def pop(self) -> str:
        if not self.isEmpty():
            self.top = -1
            return self.stack.pop()
        else:
            return ""
    
    def peek(self) -> str:
        return self.stack[-1]

    # ===== PRESEDENCE CHECKING AND OPERATOR OPERATIONS =====
    
    def isOperand(self, charector:str) -> bool:
        if charector.isalpha() or charector.isdigit():
            return True
        else:
            return False
    
    def get_precedence(self, operator:str) -> int:
        return self.precedence[operator]

    # ===== CONVERTION FUNCTIONS =====

    def infix_to_postfix(self, infix:str) -> str:

        for charector in infix :

            # charector is alphabet or a digit
            if self.isOperand(charector):
                self.output.append(charector)
            
            elif charector == "(":
                self.push(charector)

            elif charector == ")":
                # if charector == ")" pop out elements from the stack and append the elememts to the output array 
                # when while popping and appending "(" is encountered pop it and discard both the parenthesis
                stack_top_op = self.stack[-1]
                while stack_top_op != "(":
                    popped_op = self.stack.pop()
                    self.output.append(popped_op)
                    stack_top_op = self.stack[-1]
                else:
                    self.stack.pop()
            
            else:
                # this else block works when an operator (+ - * / ^) is encountered
                infix_op = charector
                while not self.isEmpty():
                    try:
                        stack_top_op = self.peek()
                    except:
                        self.top = -1
                    if stack_top_op == "(":
                        self.push(infix_op)
                        break
                    else:
                        stack_top_op_precedence = self.get_precedence(stack_top_op)
                        infix_op_precedence = self.get_precedence(infix_op)
                        if stack_top_op_precedence < infix_op_precedence:
                            self.push(infix_op)
                            break
                        else:
                            top = self.top
                            popped_op = self.pop()
                            self.output.append(popped_op)
                            self.top = top
                else:
                    self.push(infix_op)
        try:
            while True:
                popped_op = self.stack.pop()
                self.output.append(popped_op)
        except:
            pass
        
        return "".join(self.output)

if __name__ == '__main__':

    """
    test infix expresstions
    1. ((A+B)-C*(D/E))+F
    2. a+b*c+d
    3. a+b*(c^d-e)^(f+g*h)-i
    """

    infix = input("Enter infix expression : ")
    infix_obj = Convert()
    
    choise = int(input("Enter 1 for postfix and 2 for prefix :"))

    if choise == 1:
        # infix to postfix
        postfix = infix_obj.infix_to_postfix(infix)
        print(postfix)

    else:
        # infix to prefix

        # infix to prefix can be done using infix to postfix 
        # first convert "(" to ")" and ")" to "("
        # to so do reverse the in fix expression, convert it to postfix and reverese postfix to get prefix

        # replaced by "[" instead of "(" so that all brackets do not become either closing or opening brackets
        # >>> infix = "(()"
        # >>> infix.replace("(", ")")
        # ')))'
        # >>> infix = infix.replace("(", ")")
        # >>> infix
        # ')))'
        open_brackets = infix.count("(")
        close_brackes = infix.count(")")
        infix = infix.replace("(" , "[" , open_brackets) 
        infix = infix.replace(")" , "]" , close_brackes)
        infix = infix.replace("[" , ")" , open_brackets) 
        infix = infix.replace("]" , "(" , close_brackes)

        prefix = infix_obj.infix_to_postfix(infix[::-1])[::-1]
        print(prefix)

File: Stacks/infix postfix perfix converstion/convertion.c
#include <stdio.h>
#include <stdlib.h> // to use the maloc() and free() function
#include <stdbool.h> // to use the bool datatype (true and false)
#include <string.h> // for string operations (here strlen)
#include <ctype.h> // for isalpha() and isdigit() functions

// ===== OPERATIONAL VARIABLES =====

char* infix_expression;
char* output; // to store the postfix or prefix expression

char* stack; // stack used for convertion operations
int capacity;
int top = -1;

char operators[5] = "+-*/^";

// ===== STACK OPERATIONS =====

void push(int charector){
    top++;
    stack[top] = charector;
}

char pop(){
    char charector = stack[top];
    top--;
    return charector;
}

// ===== PRESEDENCE CHECKING AND OPERATOR OPERATIONS =====

int get_precedence(char operator){
    switch (operator){
        case '+':
        case '-':
            return 1;
            break;
        case '*':
        case '/':
            return 2;
            break;
        case '^':
            return 3;
            break;
        default:
            return 0;
            break;
    }
}

// fxn to find if a charector is an operand or not
bool isOperand(char charector){
    return (isdigit(charector) || isalpha(charector)) ? true : false;
}

bool isOperator(char charector){
    char* found = strchr(operators, charector);
    return (found == NULL) ? false : true;
}

// ===== CONVERSTION FUNCTIONS =====

bool check_bracket_order(char* infix, int capacity){
    int brackets_balance = 0;
    for (int position = 0; position < capacity; position++){
        char bracket = infix[position];
        if (bracket == '('){
            brackets_balance++; // symolises the number of openning brackets
        } else if (bracket == ')'){
            brackets_balance--; // neutalises an open bracket whenever a closed bracket is encountered
        }
    }
    return (brackets_balance == 0) ? true : false;
}

char* infix_to_postfix(){

    int index = 0;
    char stack_top_op;

    for (int position = 0; position < capacity; position++){
        // getting what charector to analyse in the infix expression
        char charector = infix_expression[position];
        // if charector is an operator
        if (isOperator(charector)){
            char infix_op = charector;
            stack_top_op = pop();
            while (isOperator(stack_top_op) && get_precedence(stack_top_op) >= get_precedence(infix_op)){
                output[index] = stack_top_op;
                index++;
                stack_top_op = pop();
            }
            push(stack_top_op);
            push(infix_op);
        } 
        // if charector is an operand we append it to the output expresssion
        else if(isOperand(charector)){
            output[index] = charector;
            index++;
        } 
        // if charector is an open paranthesis we push it to the stack
        else if (charector == '('){
            push(charector);
        } 
        // if charctor is a closing paranthesis we pop all operators and append it the the output expression till an open paranthesis is encountered
        else if (charector == ')'){
            char popped = pop();
            while (popped != '('){
                output[index] = popped;
                index++;
                popped = pop();
            }
        }
    }
    while (stack_top_op != '\0'){
        output[index] = stack_top_op;
        index++;
    }
    return output;
}

// ===== DRIVER CODE =====

int main(){
    
    int min_capacity = 3; // min capacity of infix expression must be 3 because there must be atleast one operator bet
    infix_expression = malloc(min_capacity * sizeof(char));
    scanf("%s", infix_expression);

    capacity  = strlen(infix_expression);
    stack = (char*) malloc(capacity * sizeof(char));
    output = (char*) malloc(capacity * sizeof(char));

    bool bracets_in_order = check_bracket_order(infix_expression, capacity);

    if (bracets_in_order){
        char* postfix = infix_to_postfix();
        printf("POSTFIX : %s", postfix);

    } else{
        printf("INVALID INFIX EXPRESSION");
    }

    free(infix_expression);
    free(stack);
    
    return 0;
}